crates/swc/tests/fixture/issue-2669/namespace-hoist-var-default/input/index.ts

```diff
@@ -0,0 +1,4 @@
+export default TEST;
+namespace TEST {
+  export const VALUE = "value";
+}
```

Comment:
My current approach cannot handle this "hosting" case.
tsc emits this:

```js
export default TEST;
var TEST;
(function (TEST) {
    TEST.VALUE = "value";
})(TEST || (TEST = {}));
```

cf. https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAE2AMwIYFcA28AqBRAZRwG4AoAO1QFtgBnMVAY2DnyLgG9S45RJY4jCOVrwAagEEAMgFU8cALxwARADdUmdMGVkAvkA

Reply:
This can be fixed by modifying visit order, but it can be done in a follow-up PR.

Reply:
Thanks for the guidance.
I wasn't sure if modifying visit order won't break other things, but let me try that in this PR.

Reply:
Addressed in this commit https://github.com/swc-project/swc/pull/2742/commits/76bcc77a4dec2da2d9dc167b84dfcdca814ec869

I'm not sure if this is what you meant by "modifying visit order", but I attempted to implement that by reordering processing order of `Vec<ModuleItem>` in `Strip::visit_mut_module_items`.
Since I'm relatively new to Rust, I had some struggle with borrow checker and I'm not sure what I wrote is an "idiomatic" way of writing things.
I try not to do `O(nlogn)` sorting and that might the reason it's looking a bit ugly.
I would appreciate if you could point out possible improvements in the code.
This change might also have some negative effect on performance since it loops module items at the top level.
Please let me know if you have some concern about that area as well.

It creates some new test failures as below.  I will take a look at these cases.

https://github.com/swc-project/swc/runs/4277636123?check_suite_focus=true#step:13:5418

```
fixture____swc_ecma_parser__tests__typescript__tsc__externalModules__es6__es6modulekindWithES5Target12__input_ts
fixture____swc_ecma_parser__tests__typescript__tsc__externalModules__esnext__esnextmodulekindWithES5Target12__input_ts
fixture____swc_ecma_parser__tests__typescript__tsc__externalModules__exportAssignmentTopLevelEnumdule__input_ts
fixture____swc_ecma_parser__tests__typescript__tsc__internalModules__DeclarationMerging__EnumAndModuleWithSameNameAndCommonRoot__input_ts
fixture____swc_ecma_parser__tests__typescript__tsc__types__objectTypeLiteral__callSignatures__callSignatureWithoutReturnTypeAnnotationInference__input_ts
```

Reply:
For the test case `es6modulekindWithES5Target12`, I found an issue of my implementation (i.e. reordering).
I put a simplified version here:

- input

```ts
export enum E {
    w = 1
}

export namespace E {
    export const y = 1;
}
```

- swc

```js
(function(E) {
    E[E["w"] = 1] = "w";
})(E1 || (E1 = {
}));
var E1;
export { E1 as E };
(function(E) {
    E.y = 1;
})(E1 || (E1 = {
}));
```

- tsc

https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBMAdgVwLZwKJwN4Cg4FwDucAvHAIy4C+uuoksciAhqsAM5gsDGwmOfIQbR4PCIg7wAnmUoBuGnSA

```js
export var E;
(function (E) {
    E[E["w"] = 1] = "w";
})(E || (E = {}));
(function (E) {
    E.y = 1;
})(E || (E = {}));
```

@kdy1 Do you have any idea how I could handle these cases?

Reply:
How about postponing emitting of 

```ts
var E1;
export { E1 as E };
```

to last?
It can be done by moving the `export var` declarations